src/main/scala/piconot/internal/PicoDSL.scala

```diff
@@ -0,0 +1,99 @@
+package piconot.internal
+
+import picolib.maze.Maze
+import picolib.semantics._
+import scalafx.application.JFXApp
+import scala.language.implicitConversions
+import scala.collection.mutable.ListBuffer
+
+object PicoDSL {
+  var numStates = 0                                          // used to keep track of how many states have been used
+  val nameToState = collection.mutable.Map[String,State]()   // used to convert state names to picolib states
+  var rules = new ListBuffer[Rule]()                         // used to store all the rules encountered thus far
+  var currentStateName:String = ""                           // a hack needed so that If statements know which state they are in
+  
+  // makes sure we have allocated a state number for the given state name
+  def addStateName(stateName: String) = {
+    if (!(nameToState contains stateName)) {
+      nameToState(stateName) = State(numStates.toString)
+      numStates = numStates + 1
+    }
+  }
+  
+  // when only a string is provided in the action field, convert it to an action
+  implicit def toAction(state: String):Action = {
+    ConcatAct(stop, state)
+  }
+  
+  // updates the current state name and processes the rules for that state
+  def WhenIn(stateName: String) = {
+    addStateName(stateName)
+    currentStateName = stateName
+    
+    ((body: Unit) => (body)) // a hack that causes the rules for the state to be processed
+  }
+  
+  // adds a new rule
+  def If(cond: Condition = defaultCondition)(action: Action):Unit = {
```

Comment:
We basically had the same idea in our code of a function If taking a condition and an action. I like your implementation here.